path/dynamic/dstarlite.go

```diff
@@ -0,0 +1,484 @@
+				u.rhs = math.Inf(1)
+				for _, t := range d.model.From(u) {
+					u.rhs = math.Min(u.rhs, d.model.Weight(d.model.Edge(u, t))+t.(*dStarLiteNode).g)
+				}
+			}
+		}
+		d.update(u)
+	}
+	d.findShortestPath()
+}
+
+func (d *DStarLite) worldNodeFor(n graph.Node) *dStarLiteNode {
+	switch w := d.model.Node(n.ID()).(type) {
+	case *dStarLiteNode:
+		return w
+	case graph.Node:
+		panic(fmt.Sprintf("D* Lite: illegal world model node type: %T", w))
+	default:
+		return &dStarLiteNode{Node: n, rhs: math.Inf(1), g: math.Inf(1), key: badKey}
+	}
+}
+
+// Here returns the current location.
+func (d *DStarLite) Here() graph.Node {
+	return d.s.Node
+}
+
+// Path returns the path from the current location to the goal and the
+// weight of the path.
+func (d *DStarLite) Path() (p []graph.Node, weight float64) {
+	u := d.s
+	p = []graph.Node{u.Node}
+	for u.ID() != d.t.ID() {
+		if math.IsInf(u.rhs, 1) {
+			return nil, math.Inf(1)
+		}
+
+		// We use stored rhs comparison to break
+		// ties between calculated rhs-coequal nodes.
+		rhsMin := math.Inf(1)
+		min := math.Inf(1)
+		var (
+			next *dStarLiteNode
+			cost float64
+		)
+		for _, _v := range d.model.From(u) {
+			v := _v.(*dStarLiteNode)
+			w := d.model.Weight(d.model.Edge(u, v))
+			if rhs := w + v.g; rhs < min || (rhs == min && v.rhs < rhsMin) {
+				next = v
+				min = rhs
+				rhsMin = v.rhs
+				cost = w
+			}
+		}
+		if next == nil {
+			return nil, math.NaN()
+		}
+		u = next
+		weight += cost
+		p = append(p, u.Node)
+	}
+	return p, weight
+}
+
+/*
+The pseudocode uses the following functions to manage the priority
+queue:
+
+      * U.Top() returns a vertex with the smallest priority of all
+        vertices in priority queue U.
+      * U.TopKey() returns the smallest priority of all vertices in
+        priority queue U. (If is empty, then U.TopKey() returns [∞;∞].)
+      * U.Pop() deletes the vertex with the smallest priority in
+        priority queue U and returns the vertex.
+      * U.Insert(s, k) inserts vertex s into priority queue with
+        priority k.
+      * U.Update(s, k) changes the priority of vertex s in priority
+        queue U to k. (It does nothing if the current priority of vertex
+        s already equals k.)
+      * Finally, U.Remove(s) removes vertex s from priority queue U.
+*/
+
+// key is a D* Lite priority queue key.
+type key [2]float64
+
+var badKey = key{math.NaN(), math.NaN()}
+
+// less returns whether k is less than other. From ISBN:0-262-51129-0 pp476-483:
+//
+//  k ≤ k' iff k₁ < k'₁ OR (k₁ == k'₁ AND k₂ ≤ k'₂)
+//
+func (k key) less(other key) bool {
+	if k != k || other != other {
+		panic("D* Lite: poisoned key")
+	}
+	return k[0] < other[0] || (k[0] == other[0] && k[1] < other[1])
+}
+
+// dStarLiteNode adds D* Lite accounting to a graph.Node.
+type dStarLiteNode struct {
+	graph.Node
+	key key
+	rhs float64
+	g   float64
+}
+
+// dStarLiteQueue is a D* Lite priority queue.
+type dStarLiteQueue struct {
+	indexOf map[int]int
```

Comment:
If you want I can go through and fix this one and the one in A* as well, but I'll do that in another PR.